pig/resources/org/partiql/pig/templates/kotlin-visitor-transform.ftl

```diff
@@ -28,43 +26,55 @@
         transformMetas(node.metas)
 
 [/#macro]
 
-
-open class VisitorTransform : DomainVisitorTransformBase() {
-    [#list domain.tuples]
+[#macro visitor_transform_class class_name source_domain dest_domain_name]
+abstract class ${class_name} : DomainVisitorTransformBase() {
+    [#list source_domain.tuples]
     //////////////////////////////////////
     // Tuple Types
     //////////////////////////////////////
-    [#items as t]
-    // Tuple ${t.kotlinName}
-    open fun transform${t.kotlinName}(node: ${t.kotlinName}): ${t.kotlinName} {
-        [@transform_fun_body domain t t.kotlinName /]
+    [#items as tuple]
+    // Tuple ${tuple.kotlinName}
+    [#if !tuple.transformAbstract]
+    open fun transform${tuple.kotlinName}(node: ${source_domain.kotlinName}.${tuple.kotlinName}): ${dest_domain_name}.${tuple.kotlinName} {
+        [@transform_fun_body dest_domain_name tuple tuple.kotlinName /]
     }
-    [@trasnform_property_functions t ""/]
+    [@transform_property_functions source_domain tuple ""/]
+    [#else]
+    abstract fun transform${tuple.kotlinName}(node:${source_domain.kotlinName}.${tuple.kotlinName}): ${dest_domain_name}.${tuple.kotlinName}
+    [/#if]
     [/#items]
     [/#list]
-    [#list domain.sums as s]
+    [#list source_domain.sums as s]
     //////////////////////////////////////
     // Sum Type: ${s.kotlinName}
     //////////////////////////////////////
-    open fun transform${s.kotlinName}(node: ${domain.kotlinName}.${s.kotlinName}) =
+    open fun transform${s.kotlinName}(node: ${source_domain.kotlinName}.${s.kotlinName}): ${dest_domain_name}.${s.kotlinName} =
         when(node) {
         [#list s.variants as v]
-            is ${domain.kotlinName}.${s.kotlinName}.${v.kotlinName} -> transform${s.kotlinName}${v.kotlinName}(node)
+            is ${source_domain.kotlinName}.${s.kotlinName}.${v.kotlinName} -> transform${s.kotlinName}${v.kotlinName}(node)
         [/#list]
         }
-[#list s.variants as t]
-    // Variant ${s.kotlinName}${t.kotlinName}
-    open fun transform${s.kotlinName}${t.kotlinName}(node: ${domain.kotlinName}.${s.kotlinName}.${t.kotlinName}): ${domain.kotlinName}.${s.kotlinName} {
-        [@transform_fun_body domain, t, "${s.kotlinName}${t.kotlinName}" /]
+[#list s.variants as tuple]
+    // Variant ${s.kotlinName}${tuple.kotlinName}
+    [#if !tuple.transformAbstract]
+    open fun transform${s.kotlinName}${tuple.kotlinName}(node: ${source_domain.kotlinName}.${s.kotlinName}.${tuple.kotlinName}): ${dest_domain_name}.${s.kotlinName}  {
+        [@transform_fun_body dest_domain_name, tuple, "${s.kotlinName}${tuple.kotlinName}" /]
     }
-
-    [@trasnform_property_functions t s.kotlinName /]
-
+    [@transform_property_functions source_domain tuple s.kotlinName /]
+    [#else]
+    abstract fun transform${s.kotlinName}${tuple.kotlinName}(node: ${source_domain.kotlinName}.${s.kotlinName}.${tuple.kotlinName}): ${dest_domain_name}.${s.kotlinName}
+    [/#if]
 [/#list]
 [/#list]
 }
+[/#macro]
```

Comment:
super nit: need a new line after final `}`. Currently if there's multiple cross-domain transforms, it will be formatted without a new line:

```kotlin
abstract class DomainAToDomainB : DomainVisitorTransformBase() {
    // ...
}
abstract class DomainCToDomainD : DomainVisitorTransformBase() {
    // ...
}
```

Reply:
actually, I've been meaning to fix that.  While it is generated code and it doesn't *strictly* *need* to be formatted correctly, I would like to keep it tidy.